From 00e5dd38f773b1b17e11c722a94ff0fb7eebe5ff Mon Sep 17 00:00:00 2001
From: Kiran Kelageri <kirankelageri@codeaurora.org>
Date: Mon, 17 Dec 2018 17:07:18 -0800
Subject: [PATCH] Bluetooth-SWB: SWB feature support.

Support for SWB codec that provides HD-Voice
quality enhancement to the existing SCo calls
is added with this change.

Change-Id: Ia21cfd2a3f0839698d595e8390979392cfc36994
---
 bta/ag/bta_ag_act.cc             | 30 ++++++++++--
 bta/ag/bta_ag_cmd.cc             | 42 +++++++++++++---
 bta/ag/bta_ag_int.h              | 18 +++++++
 bta/ag/bta_ag_main.cc            |  5 ++
 bta/ag/bta_ag_sco.cc             | 83 +++++++++++++++++++++++++++++---
 btif/Android.bp                  |  3 +-
 btif/src/bluetooth.cc            | 10 ++++
 btif/src/btif_hf.cc              | 20 ++++++++
 btif/src/btif_util.cc            |  6 +++
 device/include/esco_parameters.h | 10 ++++
 device/src/esco_parameters.cc    | 35 ++++++++++++++
 internal_include/bt_target.h     |  3 ++
 12 files changed, 249 insertions(+), 16 deletions(-)

diff --git a/bta/ag/bta_ag_act.cc b/bta/ag/bta_ag_act.cc
index 7841d0442a..e4b8a52199 100644
--- a/bta/ag/bta_ag_act.cc
+++ b/bta/ag/bta_ag_act.cc
@@ -75,6 +75,9 @@
 #include "bta_ag_twsp.h"
 #endif
 #include "device/include/device_iot_config.h"
+#if (SWB_ENABLED == TRUE)
+#include "bta_ag_swb.h"
+#endif
 
 /*****************************************************************************
  *  Constants
@@ -431,6 +434,9 @@ void bta_ag_rfc_fail(tBTA_AG_SCB* p_scb, UNUSED_ATTR tBTA_AG_DATA* p_data) {
   p_scb->peer_features = 0;
   p_scb->peer_codecs = BTA_AG_CODEC_CVSD;
   p_scb->sco_codec = BTA_AG_CODEC_CVSD;
+#if (SWB_ENABLED == TRUE)
+  p_scb->is_swb_codec = false;
+#endif
   p_scb->role = 0;
   p_scb->svc_conn = false;
   p_scb->hsp_version = HSP_VERSION_1_2;
@@ -469,6 +475,10 @@ void bta_ag_rfc_close(tBTA_AG_SCB* p_scb, UNUSED_ATTR tBTA_AG_DATA* p_data) {
   p_scb->codec_updated = false;
   p_scb->codec_fallback = false;
   p_scb->codec_msbc_settings = BTA_AG_SCO_MSBC_SETTINGS_T2;
+#if (SWB_ENABLED == TRUE)
+  p_scb->codec_swb_settings = BTA_AG_SCO_SWB_SETTINGS_Q0;
+  p_scb->is_swb_codec = false;
+#endif
   p_scb->role = 0;
   p_scb->post_sco = BTA_AG_POST_SCO_NONE;
   p_scb->svc_conn = false;
@@ -1038,12 +1048,19 @@ void bta_ag_setcodec(tBTA_AG_SCB* p_scb, tBTA_AG_DATA* p_data) {
 
   /* Check if the requested codec type is valid */
   if ((codec_type != BTA_AG_CODEC_NONE) && (codec_type != BTA_AG_CODEC_CVSD) &&
-      (codec_type != BTA_AG_CODEC_MSBC)) {
+      (codec_type != BTA_AG_CODEC_MSBC)
+#if (SWB_ENABLED == TRUE)
+      && (codec_type != BTA_AG_SCO_SWB_SETTINGS_Q0)
+#endif
+    ) {
     val.num = codec_type;
     val.hdr.status = BTA_AG_FAIL_RESOURCES;
     APPL_TRACE_ERROR("bta_ag_setcodec error: unsupported codec type %d",
                      codec_type);
     (*bta_ag_cb.p_cback)(BTA_AG_WBS_EVT, (tBTA_AG*)&val);
+#if (SWB_ENABLED == TRUE)
+    (*bta_ag_cb.p_cback)(BTA_AG_SWB_EVT, (tBTA_AG*)&val);
+#endif
     return;
   }
 
@@ -1060,8 +1077,15 @@ void bta_ag_setcodec(tBTA_AG_SCB* p_scb, tBTA_AG_DATA* p_data) {
     APPL_TRACE_ERROR("bta_ag_setcodec error: unsupported codec type %d",
                      codec_type);
   }
-
-  (*bta_ag_cb.p_cback)(BTA_AG_WBS_EVT, (tBTA_AG*)&val);
+#if (SWB_ENABLED == TRUE)
+  if (codec_type == BTA_AG_SCO_SWB_SETTINGS_Q0)
+  {
+    (*bta_ag_cb.p_cback)(BTA_AG_SWB_EVT, (tBTA_AG*)&val);
+  } else
+#endif
+  {
+    (*bta_ag_cb.p_cback)(BTA_AG_WBS_EVT, (tBTA_AG*)&val);
+  }
 }
 
 void bta_ag_handle_collision(tBTA_AG_SCB* p_scb,
diff --git a/bta/ag/bta_ag_cmd.cc b/bta/ag/bta_ag_cmd.cc
index 272215bc75..4283cfa5e1 100644
--- a/bta/ag/bta_ag_cmd.cc
+++ b/bta/ag/bta_ag_cmd.cc
@@ -74,6 +74,9 @@
 #include "bta_ag_twsp.h"
 #include "bta_ag_twsp_dev.h"
 #endif
+#if (SWB_ENABLED == TRUE)
+#include "bta_ag_swb.h"
+#endif
 
 /*****************************************************************************
  *  Constants
@@ -154,6 +157,12 @@ const tBTA_AG_AT_CMD bta_ag_hfp_cmd[] = {
      BTA_AG_AT_SET | BTA_AG_AT_READ | BTA_AG_AT_TEST, BTA_AG_AT_STR, 0, 0},
     {"+BIEV", BTA_AG_AT_BIEV_EVT, BTA_AG_AT_SET, BTA_AG_AT_STR, 0, 0},
     {"+BAC", BTA_AG_AT_BAC_EVT, BTA_AG_AT_SET, BTA_AG_AT_STR, 0, 0},
+
+#if (SWB_ENABLED == TRUE)
+    {"+%QAC", BTA_AG_AT_QAC_EVT, BTA_AG_AT_SET, BTA_AG_AT_STR, 0, 0},
+    {"+%QCS", BTA_AG_AT_QCS_EVT, BTA_AG_AT_SET, BTA_AG_AT_INT, 0,
+     BTA_AG_CMD_MAX_VAL},
+#endif
 #if (TWS_AG_ENABLED == TRUE)
     {"%QMQ", BTA_AG_TWSP_AT_QMQ_EVT, BTA_AG_AT_SET, BTA_AG_AT_INT, TWSPLUS_MIN_MIC_QUALITY, TWSPLUS_MAX_MIC_QUALITY},
     {"%QES", BTA_AG_TWSP_AT_QES_EVT, BTA_AG_AT_SET, BTA_AG_AT_INT, TWSPLUS_EB_STATE_UNKNOWN, TWSPLUS_EB_STATE_INEAR},
@@ -214,6 +223,10 @@ const tBTA_AG_RESULT bta_ag_result_tbl[] = {
     {"+CME ERROR: ", BTA_AG_LOCAL_RES_CMEE, BTA_AG_RES_FMT_INT},
     {"+BCS: ", BTA_AG_LOCAL_RES_BCS, BTA_AG_RES_FMT_INT},
     {"+BIND: ", BTA_AG_BIND_RES, BTA_AG_RES_FMT_STR},
+#if (SWB_ENABLED == TRUE)
+    {"+%QAC: ", BTA_AG_LOCAL_RES_QAC, BTA_AG_RES_FMT_STR},
+    {"+%QCS: ", BTA_AG_LOCAL_RES_QCS, BTA_AG_RES_FMT_INT},
+#endif
 #if (TWS_AG_ENABLED == TRUE)
     {"%QMQ: ", BTA_AG_TWS_QMQ_RES, BTA_AG_RES_FMT_NONE},
     {"%QES: ", BTA_AG_TWS_QES_RES, BTA_AG_RES_FMT_NONE},
@@ -1385,13 +1398,19 @@ void bta_ag_at_hfp_cback(tBTA_AG_SCB* p_scb, uint16_t cmd, uint8_t arg_type,
         p_scb->peer_codecs = bta_ag_parse_bac(p_scb, p_arg, p_end);
         p_scb->codec_updated = true;
 
-        if (p_scb->peer_codecs & BTA_AG_CODEC_MSBC) {
-          p_scb->sco_codec = UUID_CODEC_MSBC;
-          APPL_TRACE_DEBUG("Received AT+BAC, updating sco codec to MSBC");
-        } else {
-          p_scb->sco_codec = UUID_CODEC_CVSD;
-          APPL_TRACE_DEBUG("Received AT+BAC, updating sco codec to CVSD");
+#if (SWB_ENABLED == TRUE)
+        if (p_scb->is_swb_codec == false) {
+#endif
+          if (p_scb->peer_codecs & BTA_AG_CODEC_MSBC) {
+            p_scb->sco_codec = UUID_CODEC_MSBC;
+            APPL_TRACE_DEBUG("Received AT+BAC, updating sco codec to MSBC");
+          } else {
+            p_scb->sco_codec = UUID_CODEC_CVSD;
+            APPL_TRACE_DEBUG("Received AT+BAC, updating sco codec to CVSD");
+          }
+#if (SWB_ENABLED == TRUE)
         }
+#endif
         /* The above logic sets the stack preferred codec based on local and
         peer codec
         capabilities. This can be overridden by the application depending on its
@@ -1459,6 +1478,17 @@ void bta_ag_at_hfp_cback(tBTA_AG_SCB* p_scb, uint16_t cmd, uint8_t arg_type,
       bta_ag_sco_open(p_scb, NULL);
       }
       break;
+#if (SWB_ENABLED == TRUE)
+    case BTA_AG_AT_QAC_EVT:
+        p_scb->peer_codecs |= bta_ag_parse_qac(p_scb, p_arg);
+        bta_ag_swb_handle_vs_at_events(p_scb, cmd, int_arg, val);
+        bta_ag_send_ok(p_scb);
+      break;
+    case BTA_AG_AT_QCS_EVT:
+        bta_ag_send_ok(p_scb);
+        bta_ag_swb_handle_vs_at_events(p_scb, cmd, int_arg, val);
+      break;
+#endif
 #if (TWS_AG_ENABLED == TRUE)
     case BTA_AG_TWSP_AT_QMQ_EVT:
     case BTA_AG_TWSP_AT_QES_EVT:
diff --git a/bta/ag/bta_ag_int.h b/bta/ag/bta_ag_int.h
index 07271c8a3a..61ed8eee74 100644
--- a/bta/ag/bta_ag_int.h
+++ b/bta/ag/bta_ag_int.h
@@ -273,6 +273,17 @@ typedef enum {
   BTA_AG_SCO_MSBC_SETTINGS_T1,
 } tBTA_AG_SCO_MSBC_SETTINGS;
 
+#if (SWB_ENABLED == TRUE)
+typedef enum {
+  BTA_AG_SCO_SWB_SETTINGS_Q0 = 0, /* preferred/default when codec is SWB */
+  BTA_AG_SCO_SWB_SETTINGS_Q1 = 4,
+  BTA_AG_SCO_SWB_SETTINGS_Q2 = 6,
+  BTA_AG_SCO_SWB_SETTINGS_Q3 = 7,
+  BTA_AG_SCO_SWB_SETTINGS_UKNOWN = 0xFFFF,
+} tBTA_AG_SCO_SWB_SETTINGS;
+
+#endif
+
 /* type for each service control block */
 struct tBTA_AG_SCB {
   char clip[BTA_AG_AT_MAX_LEN + 1];     /* number string used for CLIP */
@@ -329,6 +340,13 @@ struct tBTA_AG_SCB {
   tBTA_AG_SCO_MSBC_SETTINGS
       codec_msbc_settings; /* settings to be used for the impending eSCO */
 
+#if (SWB_ENABLED == TRUE)
+  tBTA_AG_SCO_SWB_SETTINGS
+      codec_swb_settings; /* settings to be used for the SWB eSCO */
+
+  bool is_swb_codec;      /* Flag to determine SWB codec  */
+#endif
+
   tBTA_AG_HF_IND
       peer_hf_indicators[BTA_AG_MAX_NUM_PEER_HF_IND]; /* Peer supported
                                                   HF indicators */
diff --git a/bta/ag/bta_ag_main.cc b/bta/ag/bta_ag_main.cc
index 16f58649b2..8111efacdf 100644
--- a/bta/ag/bta_ag_main.cc
+++ b/bta/ag/bta_ag_main.cc
@@ -268,6 +268,11 @@ static tBTA_AG_SCB* bta_ag_scb_alloc(void) {
           alarm_new("bta_ag.scb_xsco_conn_collision_timer");
       /* set eSCO mSBC setting to T2 as the preferred */
       p_scb->codec_msbc_settings = BTA_AG_SCO_MSBC_SETTINGS_T2;
+#if (SWB_ENABLED == TRUE)
+      /* set eSCO SWB setting to Q0 as the preferred */
+      p_scb->codec_swb_settings = BTA_AG_SCO_SWB_SETTINGS_Q0;
+      p_scb->is_swb_codec = false;
+#endif
       APPL_TRACE_DEBUG("bta_ag_scb_alloc %d", bta_ag_scb_to_idx(p_scb));
       break;
     }
diff --git a/bta/ag/bta_ag_sco.cc b/bta/ag/bta_ag_sco.cc
index f440cec4ae..912970b456 100644
--- a/bta/ag/bta_ag_sco.cc
+++ b/bta/ag/bta_ag_sco.cc
@@ -75,6 +75,9 @@
 #include "bta_ag_twsp.h"
 #include "bta_ag_twsp_dev.h"
 #endif
+#if (SWB_ENABLED == TRUE)
+#include "bta_ag_swb.h"
+#endif
 
 #include <btcommon_interface_defs.h>
 
@@ -264,7 +267,11 @@ static void bta_ag_sco_disc_cback(uint16_t sco_idx) {
     if (is_twsp_device(curr_scb->peer_addr) != true) {
 #endif
     if (bta_ag_cb.sco.p_curr_scb != NULL &&
-           bta_ag_cb.sco.p_curr_scb->inuse_codec == BTA_AG_CODEC_MSBC) {
+           (bta_ag_cb.sco.p_curr_scb->inuse_codec == BTA_AG_CODEC_MSBC
+#if (SWB_ENABLED == TRUE)
+        || bta_ag_cb.sco.p_curr_scb->inuse_codec == BTA_AG_SCO_SWB_SETTINGS_Q0
+#endif
+       )) {
       /* Bypass vendor specific and voice settings if enhanced eSCO supported */
 
     if (!(controller_get_interface()
@@ -607,12 +614,21 @@ void bta_ag_create_sco(tBTA_AG_SCB* p_scb, bool is_orig) {
   if ((p_scb->sco_codec == BTA_AG_CODEC_MSBC) && !p_scb->codec_fallback)
     esco_codec = BTA_AG_CODEC_MSBC;
 
+#if (SWB_ENABLED == TRUE)
+  if ((p_scb->sco_codec == BTA_AG_SCO_SWB_SETTINGS_Q0) && !p_scb->codec_fallback)
+    esco_codec = BTA_AG_SCO_SWB_SETTINGS_Q0;
+#endif
+
   if (p_scb->codec_fallback) {
     p_scb->codec_fallback = false;
     /* Force AG to send +BCS for the next audio connection. */
     p_scb->codec_updated = true;
     /* Reset mSBC settings to T2 for the next audio connection */
     p_scb->codec_msbc_settings = BTA_AG_SCO_MSBC_SETTINGS_T2;
+#if (SWB_ENABLED == TRUE)
+    /* Reset SWB settings to Q3 for the next audio connection */
+    p_scb->codec_swb_settings = BTA_AG_SCO_SWB_SETTINGS_Q0;
+#endif
   }
 
   esco_codec_t codec_index = ESCO_CODEC_CVSD;
@@ -626,6 +642,19 @@ void bta_ag_create_sco(tBTA_AG_SCB* p_scb, bool is_orig) {
     }
   }
 
+#if (SWB_ENABLED == TRUE)
+  if (p_scb->is_swb_codec == true && !p_scb->codec_updated) {
+    if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q3) {
+      codec_index = ESCO_CODEC_SWB_Q3;
+    } else if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q2) {
+      codec_index = ESCO_CODEC_SWB_Q2;
+    } else if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q1) {
+      codec_index = ESCO_CODEC_SWB_Q1;
+    } else if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q0) {
+      codec_index = ESCO_CODEC_SWB_Q0;
+    }
+  }
+#endif
   /* Initialize eSCO parameters */
   enh_esco_params_t params = esco_parameters_for_codec(codec_index);
   /* For CVSD */
@@ -750,6 +779,19 @@ static void bta_ag_create_pending_sco(tBTA_AG_SCB* p_scb, bool is_local) {
 
   /* Local device requested SCO connection to peer */
   if (is_local) {
+#if (SWB_ENABLED == TRUE)
+    if (p_scb->is_swb_codec == true && !p_scb->codec_updated) {
+      if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q3) {
+        params = esco_parameters_for_codec(ESCO_CODEC_SWB_Q3);
+      } else if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q2) {
+        params = esco_parameters_for_codec(ESCO_CODEC_SWB_Q2);
+      } else if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q1) {
+        params = esco_parameters_for_codec(ESCO_CODEC_SWB_Q1);
+      } else if (p_scb->codec_swb_settings == BTA_AG_SCO_SWB_SETTINGS_Q0) {
+        params = esco_parameters_for_codec(ESCO_CODEC_SWB_Q0);
+      }
+    }
+#endif
     if (esco_codec == BTA_AG_CODEC_MSBC) {
       if (p_scb->codec_msbc_settings == BTA_AG_SCO_MSBC_SETTINGS_T2) {
         params = esco_parameters_for_codec(ESCO_CODEC_MSBC_T2);
@@ -845,9 +887,17 @@ static void bta_ag_codec_negotiation_timer_cback(void* data) {
     if (is_blacklisted == false) {
       APPL_TRACE_IMP("%s: blacklisting device %s for codec negotiation",
                     __func__, p_scb->peer_addr.ToString().c_str());
-
-      interop_database_add(INTEROP_DISABLE_CODEC_NEGOTIATION,
+#if (SWB_ENABLED == TRUE)
+      if (p_scb->is_swb_codec == false) {
+#endif
+        interop_database_add(INTEROP_DISABLE_CODEC_NEGOTIATION,
                          &p_scb->peer_addr, 3);
+#if (SWB_ENABLED == TRUE)
+      } else {
+        APPL_TRACE_IMP("%s: Ignore blacklisting SWB device  %s for codec negotiation",
+                      __func__, p_scb->peer_addr.ToString().c_str());
+      }
+#endif
     } else {
        APPL_TRACE_IMP("%s: dev %s is already blacklisted for codec negotiation",
                      __func__, p_scb->peer_addr.ToString().c_str());
@@ -895,8 +945,17 @@ void bta_ag_codec_negotiate(tBTA_AG_SCB* p_scb) {
     /* Change the power mode to Active until SCO open is completed. */
     bta_sys_busy(BTA_ID_AG, p_scb->app_id, p_scb->peer_addr);
 
-    /* Send +BCS to the peer */
-    bta_ag_send_bcs(p_scb, NULL);
+
+#if (SWB_ENABLED == TRUE)
+    if (p_scb->is_swb_codec == true && p_scb->codec_updated) {
+      /* Send +QCS to the peer */
+      bta_ag_send_qcs(p_scb, NULL);
+    } else
+#endif
+    {
+      /* Send +BCS to the peer */
+      bta_ag_send_bcs(p_scb, NULL);
+    }
 
     /* Start timer to handle timeout */
     alarm_set_on_mloop(p_scb->codec_negotiation_timer,
@@ -2255,6 +2314,11 @@ void bta_ag_sco_conn_open(tBTA_AG_SCB* p_scb,
 
     /* reset to mSBC T2 settings as the preferred */
     p_scb->codec_msbc_settings = BTA_AG_SCO_MSBC_SETTINGS_T2;
+
+#if (SWB_ENABLED == TRUE)
+    /* reset to SWB Q0 settings as the preferred */
+    p_scb->codec_swb_settings = BTA_AG_SCO_SWB_SETTINGS_Q0;
+#endif
  }
 
 /*******************************************************************************
@@ -2304,7 +2368,11 @@ void bta_ag_sco_conn_close(tBTA_AG_SCB* p_scb,
        if (p_scb->svc_conn &&
            (p_scb->codec_fallback ||
            (p_scb->sco_codec == BTM_SCO_CODEC_MSBC &&
-            p_scb->codec_msbc_settings == BTA_AG_SCO_MSBC_SETTINGS_T1))) {
+            p_scb->codec_msbc_settings == BTA_AG_SCO_MSBC_SETTINGS_T1)
+#if (SWB_ENABLED == TRUE)
+            || (p_scb->codec_fallback && p_scb->sco_codec == BTA_AG_SCO_SWB_SETTINGS_Q0)
+#endif
+        )) {
               bta_ag_sco_event(p_scb, BTA_AG_SCO_REOPEN_E);
        } else {
            bta_ag_sco_event(p_scb, BTA_AG_SCO_CONN_CLOSE_E);
@@ -2326,6 +2394,9 @@ void bta_ag_sco_conn_close(tBTA_AG_SCB* p_scb,
      /* call app callback */
      bta_ag_cback_sco(p_scb, BTA_AG_AUDIO_CLOSE_EVT);
      p_scb->codec_msbc_settings = BTA_AG_SCO_MSBC_SETTINGS_T2;
+#if (SWB_ENABLED == TRUE)
+     p_scb->codec_swb_settings = BTA_AG_SCO_SWB_SETTINGS_Q0;
+#endif
 }
 
 /*******************************************************************************
diff --git a/btif/Android.bp b/btif/Android.bp
index 665a3762bd..24c3d139a0 100644
--- a/btif/Android.bp
+++ b/btif/Android.bp
@@ -4,6 +4,7 @@ btifCommonIncludes = [
     "vendor/qcom/opensource/commonsys/system/bt",
     "vendor/qcom/opensource/commonsys/system/bt/bta/include",
     "vendor/qcom/opensource/commonsys/system/bt/bta/sys",
+    "vendor/qcom/opensource/commonsys/system/bt/bta/ag",
     "vendor/qcom/opensource/commonsys/system/bt/bta/dm",
     "vendor/qcom/opensource/commonsys/system/bt/btcore/include",
     "vendor/qcom/opensource/commonsys/system/bt/device/include",
@@ -23,7 +24,7 @@ btifCommonIncludes = [
     "vendor/qcom/opensource/commonsys/system/bt/utils/include",
     "vendor/qcom/opensource/commonsys/bluetooth_ext/system_bt_ext/btif/include",
     "vendor/qcom/opensource/commonsys/bluetooth_ext/system_bt_ext",
-
+    "vendor/qcom/opensource/commonsys/bluetooth_ext/system_bt_ext/bta/include",
     "vendor/qcom/opensource/commonsys/bluetooth/hal/include",
     "vendor/qcom/opensource/commonsys/bluetooth_ext/vhal/include",
     "vendor/qcom/opensource/commonsys-intf/bluetooth/include",
diff --git a/btif/src/bluetooth.cc b/btif/src/bluetooth.cc
index 0db847b12b..416af78041 100644
--- a/btif/src/bluetooth.cc
+++ b/btif/src/bluetooth.cc
@@ -46,6 +46,9 @@
 #include <hardware/bt_rc_ext.h>
 #include <hardware/bt_sdp.h>
 #include <hardware/bt_sock.h>
+#if (SWB_ENABLED == TRUE)
+#include <hardware/vendor_hf.h>
+#endif
 #ifdef WIPOWER_SUPPORTED
 #include <hardware/wipower.h>
 #endif
@@ -131,6 +134,9 @@ extern HearingAidInterface* btif_hearing_aid_get_interface();
 extern btvendor_interface_t *btif_vendor_get_interface();
 /* vendor socket*/
 extern btvendor_interface_t *btif_vendor_socket_get_interface();
+#if (SWB_ENABLED == TRUE)
+extern btvendor_interface_t *btif_vendor_hf_get_interface();
+#endif
 /* broadcast transmitter */
 extern ba_transmitter_interface_t *btif_bat_get_interface();
 extern btrc_vendor_ctrl_interface_t *btif_rc_vendor_ctrl_get_interface();
@@ -401,6 +407,10 @@ static const void* get_profile_interface(const char* profile_id) {
 
   if (is_profile(profile_id, BT_PROFILE_VENDOR_SOCKET_ID))
     return btif_vendor_socket_get_interface();
+#if (SWB_ENABLED == TRUE)
+  if (is_profile(profile_id, BT_PROFILE_VENDOR_HF_ID))
+    return btif_vendor_hf_get_interface();
+#endif
 
 #ifdef WIPOWER_SUPPORTED
   if (is_profile(profile_id, BT_PROFILE_WIPOWER_VENDOR_ID))
diff --git a/btif/src/btif_hf.cc b/btif/src/btif_hf.cc
index 04a23ce58b..1bd009061e 100644
--- a/btif/src/btif_hf.cc
+++ b/btif/src/btif_hf.cc
@@ -68,6 +68,10 @@ IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #include <log/log.h>
 
 #include "bta/include/bta_ag_api.h"
+#if (SWB_ENABLED == TRUE)
+#include "bta_ag_swb.h"
+#include <hardware/vendor_hf.h>
+#endif
 #include "bta/include/utl.h"
 #include "bta_ag_api.h"
 #include "btif_common.h"
@@ -804,6 +808,14 @@ static void btif_hf_upstreams_evt(uint16_t event, char* p_param) {
       }
       break;
 
+#if SWB_ENABLED
+    case BTA_AG_SWB_EVT:
+      BTIF_TRACE_DEBUG("%s: AG final selected SWB codec is 0x%02x 0=Q0 4=Q1 6=Q3 7=Q4",
+                       __func__, p_data->val.num);
+      btif_handle_vendor_hf_events(event, p_data->val.num, &btif_hf_cb[idx].connected_bda);
+      break;
+#endif
+
     /* Java needs to send OK/ERROR for these commands */
     case BTA_AG_AT_CHLD_EVT:
       HAL_HF_CBACK(bt_hf_callbacks, AtChldCallback,
@@ -862,6 +874,14 @@ static void btif_hf_upstreams_evt(uint16_t event, char* p_param) {
               &btif_hf_cb[idx].connected_bda);
       break;
 
+#if (SWB_ENABLED == TRUE)
+    case BTA_AG_AT_QCS_EVT:
+      BTIF_TRACE_DEBUG("%s: AG final selected SWB codec is 0x%02x 0=Q0 4=Q1 6=Q3 7=Q4",
+                       __func__, p_data->val.num);
+      btif_handle_vendor_hf_events(event, p_data->val.num, &btif_hf_cb[idx].connected_bda);
+      break;
+#endif
+
     case BTA_AG_AT_BIND_EVT:
       if (p_data->val.hdr.status == BTA_AG_SUCCESS) {
         HAL_HF_CBACK(bt_hf_callbacks, AtBindCallback, p_data->val.str,
diff --git a/btif/src/btif_util.cc b/btif/src/btif_util.cc
index 35d3c18bdf..ecdc091f4d 100644
--- a/btif/src/btif_util.cc
+++ b/btif/src/btif_util.cc
@@ -52,6 +52,9 @@
 #include "btif_dm.h"
 #include "btu.h"
 #include "btif_av.h"
+#if (SWB_ENABLED == TRUE)
+#include "bta_ag_swb.h"
+#endif
 
 /*******************************************************************************
  *  Constants & Macros
@@ -195,6 +198,9 @@ const char* dump_hf_event(uint16_t event) {
     CASE_RETURN_STR(BTA_AG_AT_CKPD_EVT)
     CASE_RETURN_STR(BTA_AG_DISABLE_EVT)
     CASE_RETURN_STR(BTA_AG_WBS_EVT)
+#if (SWB_ENABLED == TRUE)
+    CASE_RETURN_STR(BTA_AG_SWB_EVT)
+#endif
     CASE_RETURN_STR(BTA_AG_AT_A_EVT)
     CASE_RETURN_STR(BTA_AG_AT_D_EVT)
     CASE_RETURN_STR(BTA_AG_AT_CHLD_EVT)
diff --git a/device/include/esco_parameters.h b/device/include/esco_parameters.h
index f906c8d30e..03d14adc00 100644
--- a/device/include/esco_parameters.h
+++ b/device/include/esco_parameters.h
@@ -33,9 +33,19 @@ typedef enum {
   ESCO_CODEC_CVSD = 0,
   ESCO_CODEC_MSBC_T1,
   ESCO_CODEC_MSBC_T2,
+#if (SWB_ENABLED == TRUE)
+  ESCO_CODEC_SWB_Q0,
+  ESCO_CODEC_SWB_Q1,
+  ESCO_CODEC_SWB_Q2,
+  ESCO_CODEC_SWB_Q3,
+#endif
 } esco_codec_t;
 
+#if (SWB_ENABLED == TRUE)
+#define ESCO_NUM_CODECS 7
+#else
 #define ESCO_NUM_CODECS 3
+#endif
 
 // Coding Formats (BT 4.1 or later Assigned numbers)
 #define ESCO_CODING_FORMAT_ULAW ((uint8_t)0x00)     /* u-Law log    */
diff --git a/device/src/esco_parameters.cc b/device/src/esco_parameters.cc
index 80f5e76717..e64ea4855e 100644
--- a/device/src/esco_parameters.cc
+++ b/device/src/esco_parameters.cc
@@ -137,7 +137,42 @@ static const enh_esco_params_t default_esco_parameters[ESCO_NUM_CODECS] = {
      .packet_types =
          (ESCO_PKT_TYPES_MASK_EV3 | ESCO_PKT_TYPES_MASK_NO_3_EV3 |
           ESCO_PKT_TYPES_MASK_NO_2_EV5 | ESCO_PKT_TYPES_MASK_NO_3_EV5),
+     .retransmission_effort = ESCO_RETRANSMISSION_QUALITY},
+
+#if (SWB_ENABLED == TRUE)
+     // SWB Q0
+    {.transmit_bandwidth = TXRX_64KBITS_RATE,
+     .receive_bandwidth = TXRX_64KBITS_RATE,
+     .transmit_coding_format = {.coding_format = ESCO_CODING_FORMAT_VS,
+                                .company_id = 0x000A,
+                                .vendor_specific_codec_id = 0x0000},
+     .receive_coding_format = {.coding_format = ESCO_CODING_FORMAT_VS,
+                               .company_id = 0x000A,
+                               .vendor_specific_codec_id = 0x0000},
+     .transmit_codec_frame_size = 60,
+     .receive_codec_frame_size = 60,
+     .input_bandwidth = INPUT_OUTPUT_128K_RATE,
+     .output_bandwidth = INPUT_OUTPUT_128K_RATE,
+     .input_coding_format = {.coding_format = ESCO_CODING_FORMAT_LINEAR,
+                             .company_id = 0x0000,
+                             .vendor_specific_codec_id = 0x0000},
+     .output_coding_format = {.coding_format = ESCO_CODING_FORMAT_LINEAR,
+                              .company_id = 0x0000,
+                              .vendor_specific_codec_id = 0x0000},
+     .input_coded_data_size = 16,
+     .output_coded_data_size = 16,
+     .input_pcm_data_format = ESCO_PCM_DATA_FORMAT_2_COMP,
+     .output_pcm_data_format = ESCO_PCM_DATA_FORMAT_2_COMP,
+     .input_pcm_payload_msb_position = 0,
+     .output_pcm_payload_msb_position = 0,
+     .input_data_path = ESCO_DATA_PATH_PCM,
+     .output_data_path = ESCO_DATA_PATH_PCM,
+     .input_transport_unit_size = 0x00,
+     .output_transport_unit_size = 0x00,
+     .max_latency_ms = 14,
+     .packet_types = 0x0380,
      .retransmission_effort = ESCO_RETRANSMISSION_QUALITY}};
+#endif
 
 enh_esco_params_t esco_parameters_for_codec(esco_codec_t codec) {
   CHECK(codec >= 0) << "codec index " << (int)codec << "< 0";
diff --git a/internal_include/bt_target.h b/internal_include/bt_target.h
index 5ddcd5b01f..6897ffe79f 100644
--- a/internal_include/bt_target.h
+++ b/internal_include/bt_target.h
@@ -1407,6 +1407,9 @@ incoming connection
 #define TWS_AG_ENABLED TRUE
 #endif
 
+#ifndef SWB_ENABLED
+#define SWB_ENABLED FALSE
+#endif
 /******************************************************************************
  *
  * AVRCP
-- 
2.17.1

